### ISSUE TITLE
X509Chain validation fails on MacOS 10.15.2 using CustomTrustStore

### ISSUE THREAD

**Issue by edwardneal** [dotnet/runtime#1923]
For context, this issue looks like a root cause of a test failure in dotnet/aspnetcore#15029 (which is in turn designed to fix dotnet/aspnetcore#14718 and aspnet/AspNetCore.Docs#15456, and enable the use of non-trivial client certificate authentication on Azure App Service.) The test which is failing is [`VerifyValidClientCertWithTrustedChainAuthenticates`](https://github.com/edwardneal/AspNetCore/blob/control-client-certificate-root-validation/src/Security/Authentication/test/CertificateTests.cs#L288), but this failure is only happening on MacOS; Windows and Linux work fine.

This seems to be because `X509Chain.Build` is returning false; I'm not sure whether it's returning false because of a problem with the MacOS interface layer or because one of the certificates in the chain doesn't comply with a MacOS-specific security requirement. When it returns false though, the only status reported in `X509Chain.ChainStatus` is `PartialChain`.

I've run the same three certificates through two commands: `security verify-cert` fails with a very similar error reason to `X509Chain.ChainStatus`, and `openssl verify` succeeds. The output from `security verify-cert` is:

```
> security verify-cert -c validSignedClientCertificate.cer -r validSignedSecondaryRootCertificate.cer -r validSelfSignedPrimaryRootCertificate.cer -v -L -l

Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
---
Certificate chain
 0: Valid Signed Client
    <cert(0x7ffbcc5037e0) s: Valid Signed Client i: Valid Signed Secondary Root>
---
Certificate errors
 0: Valid Signed Client
    Unable to find next certificate in the chain [MissingIntermediate]
---
Trust evaluation results
{
    TrustEvaluationDate = "2020-01-19 14:48:01 +0000";
    TrustResultDetails =      (
                {
            MissingIntermediate = 0;
            StatusCodes =                (
                "-2147409622"
            );
        }
    );
    TrustResultValue = 5;
}
---
Trust evaluation errors
Error Domain=NSOSStatusErrorDomain Code=-25318 "errKCCreateChainFailed / errSecCreateChainFailed: / The attempt to create a certificate chain failed." UserInfo={NSLocalizedDescription=\u201cValid Signed Client\u201d certificate is not trusted, NSUnderlyingError=0x7ffbcc4064c0 {Error Domain=NSOSStatusErrorDomain Code=-25318 "errKCCreateChainFailed / errSecCreateChainFailed: / The attempt to create a certificate chain failed."
UserInfo={NSLocalizedDescription=Certificate 0 \u201cValid Signed Client\201d has errors: Unable to build chain to root (possible missing intermediate);}}}
```

I'm using a piece of sample code to reproduce the failure:
```csharp
var chain = new X509Chain();
chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
chain.ChainPolicy.CustomTrustStore.Add(Certificates.SignedSecondaryRoot);
chain.ChainPolicy.CustomTrustStore.Add(Certificates.SelfSignedPrimaryRoot);
chain.ChainPolicy.VerificationTime = Certificates.SignedClient.NotBefore.AddSeconds(1);

Asset.IsTrue(chain.Build(Certificates.SignedClient));
```

None of the certificates are being loaded with a private key, and none are in the system keychain. I've tried loading the certificates from PFX files with a private key, and I've tried adding SignedSecondaryRoot and SelfSignedPrimaryRoot to the system keychain and marking them as trusted. Neither of these have changed the outcome.

I've uploaded the three certificates [here](https://github.com/dotnet/runtime/files/4082805/Certificates.zip). I've seen the existing [CustomRootTrustDoesNotTrustIntermediates](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Security.Cryptography.X509Certificates/tests/DynamicChainTests.cs#L280) test, but can't see any major difference between the certificates generated there and these ones (besides the fact that the static ones in the ZIP file don't have any private keys - already eliminated as a factor in my tests.)

**Comment by wfurt** [comment]
If OS will not trust the chain, .NET will not either as we defer decision to Apple's API. 
Catalina has been particular tricky as Apple added new requirements for certificate to be consider valid.  (e.g. what 10.14 may consider valid , 10.15 may not) 

cc:  @bartonjs

**Comment by bartonjs** [comment]
At first I thought that it was that your root validity was too long.

Then I thought it was that Apple was enforcing the CA/Browser Forum's Baseline Requirement that root certs have no EKU.

> 7.1.2.1.Root CA Certificate
> a.basicConstraints
>
> This extension MUST appear as a critical extension.  The cA field MUST be set true.  The pathLenConstraint field SHOULD NOT be present.
>
> b.keyUsage
This extension MUST be present and MUST be marked critical.  Bit positions for keyCertSign and cRLSign MUST be set.  If the Root CA Private Key is used for signing OCSP responses, then the digitalSignature bit MUST be set.
>
> c.certificatePolicies
>
> This extension SHOULD NOT be present.
>
> d.extendedKeyUsage
>
> This extension MUST NOT be present.

Both of those were still producing valid chains, so I started going through your properties again... and the winner: Apple doesn't support RSASSA-PSS in certificate chains.  When I changed the signature type to PSS everything failed.

Since this is an OS limitation, there's nothing we can do about it.  You'll need to regenerate your certificates with RSASSA-PKCS1_v1.5.  (And, again, I have to recommend that for unit tests you use generated certificates, so you can a) more easily observe the actual changes to your tests and b) avoid long-lifetime certificates while also avoiding frequent checkins due to the clock advancing).

**Comment by edwardneal** [comment]
I've just regenerated my sample certificates and retested, and I've replicated your successful result. In my particular case, I was using the `New-SelfSignedCertificate` cmdlet with the `-AlternateSignatureAlgorithm` parameter - copied from another piece of code which builds up a certificate hierarchy. Thanks for your help.

I wanted to keep my original PR pretty small and kept my new unit tests consistent with the ones which were already there, but I agree that generating certificates dynamically would be the right way to work. I'll open a second PR later which does that so that the current one stays on topic.
